x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/core/security/transport/netty4/SecurityNetty4Transport.java

```diff
@@ -159,7 +166,21 @@ protected InboundPipeline getInboundPipeline(boolean isRemoteClusterServerChanne
                 : new InboundDecoder(recycler),
             new InboundAggregator(getInflightBreaker(), getRequestHandlers()::getHandler, ignoreDeserializationErrors()),
             this::inboundMessage
-        );
+        ) {
+            @Override
+            protected void headerReceived(Header header) {
+                if (isRemoteClusterServerChannel && header.isRequest() && header.isHandshake() == false) {
+                    assert header.needsToReadVariableHeader() == false;
+                    // start authn asynchronously
+                    // TODO maybe stash the thread context
+                    crossClusterAccessAuthenticationService.tryAuthenticate(header.getRequestHeaders(), ActionListener.wrap(aVoid -> {
+                        // NOOP
+                    }, e -> uncaughtExceptionReference.set(new HeaderValidationException(header, e))));
+                }
+                // go on with the message parts
+                super.headerReceived(header);
+            }
```

Comment:
Sorry to bringing this up again. But I find myself leaning more towards a more explicit separation between RCS server channel and others. Can we have something like the follows for the `getInboundPipeline` method?

```java
@Override
protected InboundPipeline getInboundPipeline(boolean isRemoteClusterServerChannel) {
    if (false == isRemoteClusterServerChannel) {
        return super.getInboundPipeline(isRemoteClusterServerChannel);
    } else {
        return new InboundPipeline(...) {
             @Override
              protected void headerReceived(Header header) {
                  ...
              }
        } 
    }
}
```

This way we avoid running any additional logic (even if they are just simple `if`s) in channels other than RCS 2.0 server.

Reply:
I've pushed [6360029](https://github.com/elastic/elasticsearch/pull/98814/commits/636002984c5a5939aa4cdd9f32abd80c6ff79e2e)